algo-builder-tester/test/v6trade-test.ts

```diff
@@ -0,0 +1,401 @@
+      const SCCreator = runtime.getApp(appID).creator;
+      assert.equal(btcCreator.address, admin.address);
+      assert.equal(usdCreator.address, admin.address);
+      assert.equal(SCCreator, admin.address);
+    });
+    it("assert alice opted into ASA", function () {
+      syncAccounts();
+      const last_msg: Uint8Array = alice.getLocalState(
+        appID,
+        "last_msg"
+      ) as Uint8Array;
+      assert.equal(u8a2Str(last_msg), "OptIn OK.");
+    });
+  });
+  describe("buy/sell with smart contract, static price", function () {
+    let aliceCallParam: typesW.AppCallsParam;
+    let alicePayTxParam: typesW.AssetTransferParam;
+    let aliceCollectTxParam: typesW.AssetTransferParam;
+    this.beforeEach(function () {
+      aliceCallParam = {
+        type: typesW.TransactionType.CallApp,
+        sign: typesW.SignType.SecretKey,
+        fromAccount: alice.account,
+        appID: appID,
+        payFlags: { totalFee: fee },
+        appArgs: [],
+      };
+      alicePayTxParam = {
+        type: typesW.TransactionType.TransferAsset,
+        sign: typesW.SignType.SecretKey,
+        assetID: 0,
+        amount: 0,
+        fromAccount: alice.account,
+        toAccountAddr: admin.address,
+        payFlags: { totalFee: fee },
+      };
+      aliceCollectTxParam = {
+        type: typesW.TransactionType.TransferAsset,
+        sign: typesW.SignType.SecretKey,
+        assetID: 0,
+        amount: 0,
+        fromAccount: admin.account,
+        toAccountAddr: alice.address,
+        payFlags: { totalFee: fee },
+      };
+    });
+
+    function resetInitStatus() {
+      // TODO:: can't "beforeEach" : new describe()
+      syncAccounts();
+      function setAccountAsset(
+        account: AccountStore,
+        assetID: number,
+        amount: bigint
+      ) {
+        return (runtime.getAccount(account.address).assets.get(assetID)![
+          "amount"
+        ]! = amount);
+      }
+      setAccountAsset(alice, btcID, dispensedInit);
+      setAccountAsset(alice, usdID, dispensedInit);
+      setAccountAsset(billy, btcID, dispensedInit);
+      setAccountAsset(billy, usdID, dispensedInit);
+      setAccountAsset(admin, btcID, initialAdminBtc);
+      setAccountAsset(admin, usdID, initialAdminUsd);
+
+      assert.equal(dispensedInit, alice.assets.get(usdID)!["amount"]!); // 1k from dispense
+      assert.equal(dispensedInit, alice.assets.get(btcID)!["amount"]!); // 1k from dispense
+      assert.equal(initialAdminUsd, admin.assets.get(usdID)!["amount"]!); // 2k dispensed
+      assert.equal(initialAdminBtc, admin.assets.get(btcID)!["amount"]!); // 2k dispensed
+      assert.equal(0n, alice.getLocalState(appID, "AAA_balance")); // holding 0 AAA
+    }
+    this.beforeEach(resetInitStatus);
+    this.afterEach(resetInitStatus);
+    it("buy aBTC of 2aUSD ", function () {
+      const aUsdPaid = BigInt(2e6); // 2aUSD,
+      const aBtcCollected = BigInt(
+        Math.floor((Number(aUsdPaid) * 1e8) / 1e6 / 38613.14)
```

Comment:
The test semantic here is problematic. You know that float calculation is inprecise. And what actually calculated in bitcoin-approval.teal is integer division, which is different than float division here. I think you can convert them to explicit JavaScript BigInt, and doing integer division.
Samely, the price should also be represented as the same form (int) in teal, not float.

Reply:
ah i see it's generated from trader contract py

Reply:
No, it's better to replicate the exact logic of calculation from pyteal in JS, not just put the calculation in test. We can verify the number if algorithm change and loop to test a couple of other prices.
The test should be something like:
```
// in config file, also assume it's stored as a uint, not float, because when we read a price from another contract in future, it's an algorand uint64
const btcPrice = BigInt(38163140000)
const aBtcCollected = BigInt(aUsdPaid) / btcPrice
```

Reply:
During the commit I remembered my intention of writing this is applying a "real world" algorithm that we would normally write, and make sure the contract has the same effect as that. I'm thinking this might be better? @bo-abstrlabs

Reply:
resolved by https://github.com/Artcoin-Network/artificial-algorand-contract/commit/2e3e79acf41414b6cdff984ecfb6c75fc18498ed